File: reference/python/yawlabs_awsp/_replay.py
"""Replay-protection storage hook.

The Receiver MUST keep a store of recently-seen nonces. On each verification,
the store is asked: "was this nonce seen recently?" If yes, the request is
rejected as replayed; if no, the nonce is recorded with a TTL.

`InMemoryReplayStore` is a reference implementation suitable for tests and
single-process receivers. Production multi-replica deployments should use a
shared store (Redis SET NX EX, Memcached add, etc.) so a nonce seen on
replica A is also rejected on replica B.
"""

from __future__ import annotations

import time
from typing import Callable, Protocol, runtime_checkable


@runtime_checkable
class ReplayStore(Protocol):
    """Replay-protection storage hook contract.

    Implementations atomically record-or-reject. Returning True means the
    nonce was unseen and has now been recorded with TTL = ttl_seconds.
    Returning False means the nonce was already seen within its TTL window
    and the caller MUST reject the request as replayed.
    """

    def check_and_store(self, nonce: str, ttl_seconds: int) -> bool:
        """Atomically: if `nonce` has been seen within `ttl_seconds`, return
        False. Otherwise record `nonce` with TTL = `ttl_seconds` and return
        True.
        """
        ...


class InMemoryReplayStore:
    """Reference in-memory replay store.

    Suitable for tests and single-process receivers; production multi-replica
    deployments should use Redis (SET NX EX) or equivalent so replay state is
    shared.

    Memory bound: in steady state, the store holds at most one entry per
    in-flight nonce. Eviction sweeps run at most every 256 check_and_store
    calls past the soft threshold of 8192 entries, so worst-case occupancy
    between sweeps is roughly threshold + (256 * peak insert rate per call)
    -- in practice well under 9000 nonces. The threshold trades a slightly
    larger working set for amortized O(1) eviction cost: the previous
    "sweep every call past threshold" strategy was O(n) per call once
    crossed, pathological under load.
    """

    # Soft cap. Below this, no sweeping. Above this, sweep every
    # _EVICT_INTERVAL calls. Sized for low-traffic receivers; production
    # multi-replica deployments use Redis where this code path doesn't run.
    _EVICT_THRESHOLD = 8192
    _EVICT_INTERVAL = 256

    def __init__(self, clock: Callable[[], int] | None = None) -> None:
        self._clock: Callable[[], int] = clock or (lambda: int(time.time()))
        self._seen: dict[str, int] = {}
        self._calls: int = 0

    def check_and_store(self, nonce: str, ttl_seconds: int) -> bool:
        self._calls += 1
        now = self._clock()
        self._maybe_evict(now)
        expires_at = self._seen.get(nonce)
        if expires_at is not None and expires_at > now:
            return False
        self._seen[nonce] = now + ttl_seconds
        return True

    def _maybe_evict(self, now: int) -> None:
        # Amortized O(1): the O(n) sweep happens once every 256 calls, and
        # only after the dict has grown past _EVICT_THRESHOLD. n is bounded
        # by the time window since entries past their TTL get removed on
        # each sweep. For higher-throughput use, swap to a TTL-aware data
        # structure (heap, sorted set, etc.).
        if (
            len(self._seen) > self._EVICT_THRESHOLD
            and self._calls % self._EVICT_INTERVAL == 0
        ):
            expired = [k for k, v in self._seen.items() if v <= now]
            for k in expired:
                del self._seen[k]

File: reference/go/replay.go
// AWSP replay storage hook.
//
// AWSP requires the receiver to reject Deliveries whose nonce was already
// seen within the replay window (plus a 60s buffer). The storage layer is
// pluggable: production multi-replica receivers should back this with Redis
// (`SET key NX EX <ttl>`) or an equivalent atomic data store. The
// InMemoryReplayStore here is suitable for tests and single-process
// receivers.

package awsp

import (
	"sync"
	"time"
)

// ReplayStore is the interface AWSP uses to enforce nonce uniqueness.
//
// CheckAndStore atomically:
//
//   - If nonce has been recorded for configID within the past ttlSeconds,
//     return (false, nil).
//   - Otherwise record it (with TTL = ttlSeconds) and return (true, nil).
//
// configID lets a single store partition replay state per webhook
// configuration. Receivers that share state across configurations may pass
// the empty string -- nonce uniqueness is global to the receiver per spec
// section 7.2, but operators sometimes want a tighter scope to bound
// storage growth. Both behaviors satisfy the spec.
//
// Implementations MUST be safe for concurrent use across goroutines.
type ReplayStore interface {
	CheckAndStore(configID string, nonce []byte, ttlSeconds int) (firstSeen bool, err error)
}

// evictThreshold is the map size above which we may sweep. Below this,
// CheckAndStore is O(1). Above it, we sweep at most every evictStride
// calls -- the amortized cost per call stays O(1) because n grows
// proportionally to the steady-state arrival rate within the TTL window.
//
// Bound: in steady state, len(seen) is at most one TTL-window's worth
// of arrivals. With the default replayWindow=300s and a 60s buffer
// (ttl=360s), a Receiver doing 100 verifies/sec sees ~36000 entries
// at peak before eviction kicks in -- well above evictThreshold, so
// sweeps run on every evictStride'th call. At 10 verifies/sec, peak
// is ~3600 -- below threshold, no sweep needed.
const (
	evictThreshold = 8192
	evictStride    = 256
)

// InMemoryReplayStore is a simple ReplayStore for tests and single-replica
// receivers. It is safe for concurrent use.
//
// Production multi-replica deployments should swap this for a Redis-backed
// store; nonce state held in process memory does not survive restarts and
// is not shared across pods.
type InMemoryReplayStore struct {
	mu        sync.Mutex
	seen      map[string]time.Time
	clock     func() time.Time
	callCount uint64 // call counter for stride-based eviction; mod evictStride
}

// NewInMemoryReplayStore returns a fresh in-memory ReplayStore using the
// real wall clock.
func NewInMemoryReplayStore() *InMemoryReplayStore {
	return &InMemoryReplayStore{
		seen:  make(map[string]time.Time),
		clock: time.Now,
	}
}

// newInMemoryReplayStoreWithClock is an internal helper for tests that
// need deterministic time.
func newInMemoryReplayStoreWithClock(clock func() time.Time) *InMemoryReplayStore {
	return &InMemoryReplayStore{
		seen:  make(map[string]time.Time),
		clock: clock,
	}
}

// CheckAndStore implements ReplayStore.
func (s *InMemoryReplayStore) CheckAndStore(configID string, nonce []byte, ttlSeconds int) (bool, error) {
	now := s.clock()
	key := configID + ":" + string(nonce)

	s.mu.Lock()
	defer s.mu.Unlock()

	s.callCount++
	s.evictLocked(now)

	if expiresAt, ok := s.seen[key]; ok && expiresAt.After(now) {
		return false, nil
	}

	s.seen[key] = now.Add(time.Duration(ttlSeconds) * time.Second)
	return true, nil
}

// evictLocked sweeps expired entries from the map. To keep CheckAndStore
// amortized O(1), we only sweep every evictStride calls once the map has
// grown past evictThreshold. The callCount modulo gate avoids the
// pathological "sweep on every call past threshold" pattern under
// sustained high load. Callers must hold s.mu.
func (s *InMemoryReplayStore) evictLocked(now time.Time) {
	if len(s.seen) <= evictThreshold {
		return
	}
	if s.callCount%evictStride != 0 {
		return
	}
	for k, exp := range s.seen {
		if !exp.After(now) {
			delete(s.seen, k)
		}
	}
}
